src/jit/simdcodegenxarch.cpp

```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
+    regNumber      tmpReg;
+    regNumber      tmpReg2;
+    InstructionSet iset = compiler->getSIMDInstructionSet();
+
+    if (iset == InstructionSet_AVX || (baseType == TYP_ULONG))
+    {
+        // For AVX, we need two xmm tmpRegs.
+        regMaskTP tmpRegsMask = (simdNode->gtRsvdRegs & RBM_ALLFLOAT);
+        assert(tmpRegsMask != RBM_NONE);
+        assert(genCountBits(tmpRegsMask) == 2);
+
+        regMaskTP tmpReg1Mask = genFindLowestBit(tmpRegsMask);
+        tmpRegsMask &= ~tmpReg1Mask;
+        tmpReg  = genRegNumFromMask(tmpReg1Mask);
+        tmpReg2 = genRegNumFromMask(tmpRegsMask);
+    }
+    else
+    {
+        if ((intrinsicID == SIMDIntrinsicConvertToDouble) && (targetReg != op1Reg))
+        {
+            // We don't need a tmpReg; we can just use the targetReg.
+            tmpReg = targetReg;
+        }
+        else
+        {
+            tmpReg = genRegNumFromMask(simdNode->gtRsvdRegs & RBM_ALLFLOAT);
+            assert(tmpReg != op1Reg);
+        }
+        tmpReg2 = REG_NA;
+    }
+
+    if ((intrinsicID == SIMDIntrinsicConvertToDouble) && (baseType == TYP_ULONG))
+    {
+        inst_RV_RV(INS_movdqu, tmpReg2, op1Reg, baseType, emitActualTypeSize(simdType));
+        if (targetReg != op1Reg)
+        {
+            inst_RV_RV(INS_movdqu, targetReg, op1Reg, baseType, emitActualTypeSize(simdType));
+        }
+        getEmitter()->emitIns_R_I(INS_psrlq, emitActualTypeSize(simdType), targetReg, 32);
+        getEmitter()->emitIns_R_I(INS_psllq, emitActualTypeSize(simdType), tmpReg2, 32);
+        getEmitter()->emitIns_R_I(INS_psrlq, emitActualTypeSize(simdType), tmpReg2, 32);
+        getEmitter()->emitIns_R_I(INS_mov, EA_8BYTE, tmpIntReg, 0X4530000000000000);
+        inst_RV_RV(INS_mov_i2xmm, tmpReg, tmpIntReg, TYP_ULONG);
+        if (iset == InstructionSet_AVX)
+        {
+            inst_RV_RV(INS_vpbroadcastq, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        else
+        {
+            inst_RV_RV(INS_movlhps, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        inst_RV_RV(INS_orpd, targetReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        inst_RV_RV(INS_subpd, targetReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        getEmitter()->emitIns_R_I(INS_mov, EA_8BYTE, tmpIntReg, 0X4330000000000000);
+        inst_RV_RV(INS_mov_i2xmm, tmpReg, tmpIntReg, TYP_ULONG);
+        if (iset == InstructionSet_AVX)
+        {
+            inst_RV_RV(INS_vpbroadcastq, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        else
+        {
+            inst_RV_RV(INS_movlhps, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        inst_RV_RV(INS_orpd, tmpReg2, tmpReg, simdType, emitActualTypeSize(simdType));
+        inst_RV_RV(INS_subpd, tmpReg2, tmpReg, simdType, emitActualTypeSize(simdType));
+        inst_RV_RV(INS_addpd, targetReg, tmpReg2, simdType, emitActualTypeSize(simdType));
```

Comment:
Please add comments to describe at a high-level what this logic is achieving required conversion.